### ISSUE TITLE
Overlaying Objects and depth_range

### ISSUE THREAD

**Issue by sharon92** [pygfx/pygfx#585]
Overlayed opaque object doesn't display properly when zoomed out.

zoomed out:
![image](https://github.com/pygfx/pygfx/assets/51754340/8d84a00e-097a-454c-8974-b6e940c1d061)

zoomed in:
![image](https://github.com/pygfx/pygfx/assets/51754340/e9659de4-6579-4b4d-9eba-0053bdf1d146)

![image](https://github.com/pygfx/pygfx/assets/51754340/72419811-6efb-47f4-9725-b06a4a5057d2)

I have played a bit with depth_range but I can't get it to render properly when zoomed out.

Here is how I am basically drawing this:

```
        self._renderer = gfx.WgpuRenderer(
            self._canvas, blend_mode="weighted_plus"  # , sort_objects=True
        )
        self._camera = gfx.PerspectiveCamera(depth_range=(0.01, 100000))
        self._controller = gfx.OrbitController(camera=self._camera, register_events=self._renderer)

         # elevation
         elev = gfx.Mesh(
                gfx.Geometry(
                    positions=patches.geometry.positions,
                    indices=patches.geometry.indices,
                    texcoords=self.obj[self.aktiv]["texcoords"],
                    colors=self.obj[self.aktiv]["elevColor"],
                ),
                gfx.MeshBasicMaterial(
                    vertex_colors=self.elevSolid.isChecked(),
                    map=self.customGradient(
                        self.elevGradientPick.currentText()
                    ),
                    side="FRONT"
                ),
                render_mask="opaque",
            )
        self._scene.add(elev)

        # water level
         water = gfx.Mesh(
                    gfx.Geometry(
                        positions=pos,
                        indices=self.obj[self.aktiv][
                            "patches"
                        ].geometry.indices,
                        colors=color,
                        texcoords=texcoords,
                    ),
                    gfx.MeshBasicMaterial(
                        vertex_colors=self.dsSolid.isChecked(),
                        flat_shading=True,
                        map=self.customGradient(
                            self.dsGradientPick.currentText()
                        ),
                        side="FRONT",
                        clipping_mode="all",
                        # opacity=0.95
                    ),
                    render_mask="opaque",
                    render_order=0,
                )
         self._scene.add(water)

```

**Comment by sharon92** [comment]
@Korijn yep I tried this by adding 5mm -1cm to the Z Value with no difference. Also the water level is already higher than the surface geometry.

**Comment by panxinmiao** [comment]
This is due to the floating-point precision issue in the depth buffer. The depth buffer's Z value in view space is non-linear, meaning that the closer it gets to the far plane, the worse the precision becomes. To improve the precision of objects close to the far plane in the Z-buffer, you should try to minimize the ratio of far/near values for the perspective camera (ideally below 500). For example, you can set the depth range to (500, 100000).

Additionally, to achieve better visual results, when moving the camera, you may need to dynamically adjust the near and far planes based on the camera's height.

**Comment by almarklein** [comment]
Yes. To add to what @panxinmiao said, the precision falls off exponentially, with most precision being close to the near plane. You can probably set your far plane to a big value that will make the hole scene is easily visible from all angles. Then it may be enough to only set the near plane. I like the idea to automatically set it based on the camera's height.

**Comment by sharon92** [comment]
@panxinmiao @almarklein thanks for the tips!

Dynamically adjusting the depth_range works well. 

Along with the camera height, I also accounted for elevation in the calculation. 
This helps when I am viewing the scene from a lower angle.

```
        up = self._camera.get_state()["reference_up"]
        forward = la.vec_transform_quat((0, 0, -1), self._camera.get_state()["rotation"])
        elevation = la.vec_angle(forward, up) - 0.5 * np.pi
        
        height = self._camera.get_state()['height']
        near = max(0.01,abs(height*elevation/10))
       
        self._camera.depth_range = near,1e5
```

It's not perfect yet, I am still encountering angles where my scene gets chopped off. 
Working right now on fine-tuning the dynamic adjustment.

Maybe this calculation could be done for the automatic calculation of depth_range in the camera?

**Comment by almarklein** [comment]
> Maybe this calculation could be done for the automatic calculation of depth_range in the camera?

The problem with that is that the best values depend heavily on the scene. In your scene the data is very "flat" so you can use tricks like using camera height and elevation, but this is not always the case.

**Comment by Korijn** [comment]
You also might be able to leverage `self._camera.world.up` and `self._camera.world.forward` in your code to simplify it a little bit

**Comment by almarklein** [comment]
The camera has recently got some improvements to have more sensible depth range by default, and setting `depth_range` is also still possible to solve cases like this.
